matcaffe VS project and scripts added.

lunzueta opened the pull request:
This pull request is related to issue #19.

I created a VS project and the corresponding scripts, in a similar way to those for pycaffe, to be able to compile matcaffe with VS2013. For that I also took into account a couple of things from (https://github.com/happynear/caffe-windows) to export the mex file correctly. With this compilation I could run the classification_demo.m successfully in Matlab 2014b.

zer0n commented on windows/CommonSettings.props.example:
```diff
@@ -11,6 +11,10 @@
          set to the root of your Python installation. If your Python installation
          does not contain debug libraries, debug build will not work. -->
         <PythonSupport>false</PythonSupport>
+		<!-- NOTE: If Matlab support is enabled, MatlabDir (below) needs to be
```
[Misalignment] Could you please convert tabs to spaces?

lunzueta replied:
Sure. I'll do that in a moment.

zer0n commented on windows/CommonSettings.props.example:
```diff
@@ -11,6 +11,10 @@
          does not contain debug libraries, debug build will not work. -->
         <PythonSupport>false</PythonSupport>
+		<!-- NOTE: If Matlab support is enabled, MatlabDir (below) needs to be
+         set to the root of your Matlab installation, and Python support has to
```
Why does Python support need to be disabled when Matlab is enabled?

lunzueta replied:
Well, initially I tried to do it in a way that didn't matter if pycaffe support was enabled or not, but then while managing the dependencies and the linking process I found a bit strange having to add Python dependencies to this compilation. I think that it's better to remove them.

aaronzs commented on windows/scripts/MatlabPostBuild.cmd:
```diff
@@ -0,0 +1,13 @@
+copy /y "%SOLUTION_DIR%..\matlab\demo\*.m" "%OUTPUT_DIR%matcaffe\demo"
+copy /y "%SOLUTION_DIR%..\matlab\hdf5creation\*.m" "%OUTPUT_DIR%matcaffe\hdf5creation"
+move /y "%OUTPUT_DIR%caffe_.*" "%OUTPUT_DIR%matcaffe\+caffe\private"
+copy /y "%OUTPUT_DIR%*.dll" "%OUTPUT_DIR%matcaffe\+caffe\private"
```
It is not necessary to copy dlls to `+caffe/private`, because MATLAB still can't find them. These dlls need to be added to system/user PATH or put in MATLAB working directory. Am I right?

lunzueta replied:
Yes, you are right. It's not necessary to copy the dlls if they are in the system/user PATH. I just removed that copy post-build process to +caffe/private.

pavlejosipovic commented on windows/CommonSettings.props.example:
```diff
@@ -11,6 +11,10 @@
         <PythonSupport>false</PythonSupport>
+        <!-- NOTE: If Matlab support is enabled, MatlabDir (below) needs to be
+         set to the root of your Matlab installation, and Python support has to
+         be disabled. Only works in Release mode. -->
```
why python needs to be disabled when matlab is enabled?
why does it work only in debug? Is this just c/p from python setup?

aaronzs replied:
@pavlejosipovic @lunzueta I'm also working on matcaffe. I can fix this and send PR to lunzueta's repo.

lunzueta replied:
If PythonSupport is enabled then the libboost_python-vc120-mt-1_59.lib dependency requirement is transferred to the matcaffe compilation. Then, I get this linking error (LINK : fatal error LNK1104: cannot open file 'libboost_python-vc120-mt-1_59.lib'), which probably can be solved with some conditions in the configuration file, but anyway, the functionalities from that dependence won't be used in practice in Matlab, so the easiest solution is just to disable PythonSupport when MatlabSupport is enabled.

Regarding the release/debug issue, I think there's a misunderstanding here. It works only in release, but not in debug. It is possible to compile this matcaffe project in debug with the Matlab libraries (libmx.lib;libmex.lib;libmat.lib) as dependencies, but then when testing the generated mex file in Matlab, it says "invalid MEX-file", which doesn't happen when compiled in release mode. Not sure why, but I guess this happens because those Matlab libraries are only available in release mode and there might be some kind of incompatibility with the rest of dependencies because of that.

pavlejosipovic commented:
Can you add one small section on README.md on how to use MatLab wrapper?

Please squash all commits into one, once you are done addressing feedback from PR.

Thanks for helping out!

lunzueta commented:
@pavlejosipovic Sure, no problem. I'll add the section on README.md once we clarify all the issues with this PR.

pavlejosipovic commented on windows/matcaffe/matcaffe.vcxproj:
```diff
@@ -0,0 +1,133 @@
+    <Error Condition="!Exists('..\..\..\NugetPackages\boost_filesystem-vc120.1.59.0.0\build\native\boost_filesystem-vc120.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\..\..\NugetPackages\boost_filesystem-vc120.1.59.0.0\build\native\boost_filesystem-vc120.targets'))" />
+    <Error Condition="!Exists('..\..\..\NugetPackages\boost_system-vc120.1.59.0.0\build\native\boost_system-vc120.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\..\..\NugetPackages\boost_system-vc120.1.59.0.0\build\native\boost_system-vc120.targets'))" />
+    <Error Condition="!Exists('..\..\..\NugetPackages\boost_thread-vc120.1.59.0.0\build\native\boost_thread-vc120.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\..\..\NugetPackages\boost_thread-vc120.1.59.0.0\build\native\boost_thread-vc120.targets'))" />
+    <Error Condition="!Exists('..\..\..\NugetPackages\boost_python2.7-vc120.1.59.0.0\build\native\boost_python-vc120.targets')" Text="$([System.String]::Format('$(ErrorText)', '..\..\..\NugetPackages\boost_python2.7-vc120.1.59.0.0\build\native\boost_python-vc120.targets'))" />
```
I think that boost_python is only necessary for python project, you should be able to remove it from this project

lunzueta replied:
OK. I didn't realize this dependence was there too. I removed it in from my local copy (which I'll upload in a moment). Then, just in case, I tested to compile everything with both PythonSupport and MatlabSupport activated, but matcaffe still needs to link boost_python, if so. I guess it is because a dependence uses it and forces matcaffe to link it. Thus, it seems we still need to deactivate PythonSupport if we want to compile matcaffe.

By the way, this time I tested the mex file in Matlab 2015b and it also works. So, up to now the tested Matlab versions have been: 2014b, 2015a and 2015b.

aaronzs replied:
It was added by me... Because I had linker error on `libboost_python-vc120-mt-1_59.lib` yesterday. I removed it and tested it today. It is OK to remove and I've pushed my changes.

lunzueta replied:
@aaronzs, yes, I pulled your changes and merged with mine. Everything is fine in the last commit.

mnidza commented on windows/matcaffe/matcaffe.vcxproj:
```diff
@@ -0,0 +1,133 @@
+    <Import Project="..\..\..\NugetPackages\boost_filesystem-vc120.1.59.0.0\build\native\boost_filesystem-vc120.targets" Condition="Exists('..\..\..\NugetPackages\boost_filesystem-vc120.1.59.0.0\build\native\boost_filesystem-vc120.targets')" />
+    <Import Project="..\..\..\NugetPackages\boost_system-vc120.1.59.0.0\build\native\boost_system-vc120.targets" Condition="Exists('..\..\..\NugetPackages\boost_system-vc120.1.59.0.0\build\native\boost_system-vc120.targets')" />
+    <Import Project="..\..\..\NugetPackages\boost_thread-vc120.1.59.0.0\build\native\boost_thread-vc120.targets" Condition="Exists('..\..\..\NugetPackages\boost_thread-vc120.1.59.0.0\build\native\boost_thread-vc120.targets')" />
+    <Import Project="..\..\..\NugetPackages\boost_python2.7-vc120.1.59.0.0\build\native\boost_python-vc120.targets" Condition="Exists('..\..\..\NugetPackages\boost_python2.7-vc120.1.59.0.0\build\native\boost_python-vc120.targets')" />
```
I don't think matcaffe needs boost-python or protoc.

mnidza commented on windows/matcaffe/packages.config:
```diff
@@ -0,0 +1,18 @@
+  <package id="OpenBLAS" version="0.2.14.1" targetFramework="Native" />
+  <package id="OpenCV" version="2.4.10" targetFramework="Native" />
+  <package id="protobuf-v120" version="2.6.1" targetFramework="Native" />
+  <package id="protoc_x64" version="2.6.1" targetFramework="Native" />
```
I think that protoc is also not needed.

lunzueta replied:
OK. I'll remove it. Thanks.

mnidza commented:
I think that boost-python needs to put back into matcaffe. I tried building matcaffe both Python and Matlab enabled in CommonSettings.props, and got a link error.
Other projects unrelated to pycaffe (e.g. convert_imageset, compute_image_mean etc.) also have this dependency for the same reason.

lunzueta commented:
I added the python_boost dependency to matcaffe and now it works also with PythonSupport activated (I don't understand why I couldn't before). I've updated README.md too.

mnidza commented:
Can you try running the unit tests by invoking caffe.run_tests()?  That should work in Matlab 2015a and 2015b, but for me it does not for some reason.

lunzueta commented:
@mnidza I just tried it in Matlab 2014b and 2015b, with the change you mention for test_solver.m, and it works fine in both. Otherwise, I got an error too.

aaronzs commented:
@mnidza @lunzueta I tried in Matlab 2015a, change to `strrep(model_file, '\', '/')` can get rid of crash but change to `strrep(model_file, '\', '\\')` will still crash, right?

lunzueta commented:
@aaronzs yes, that's right. With `strrep(model_file, '\', '\\')` it crashed but not with `strrep(model_file, '\', '/')`

mnidza commented:
@aaronzs @lunzueta Sorry, I forgot that backslashes in `fprintf'`s format string need to be escaped as well, so `strrep(model_file, '\', '\\\\')` should work. However, I agree that the first option (replacing by forward slashes) is better because it works in Linux too.

mnidza commented on windows/CommonSettings.props.example:
```diff
@@ -64,6 +72,11 @@
     </ItemDefinitionGroup>
+    <ItemDefinitionGroup Condition="'$(MatlabSupport)'=='true'">
+        <ClCompile>
+            <PreprocessorDefinitions>MATLAB_MEX_FILE;%(PreprocessorDefinitions)</PreprocessorDefinitions>
```
Is `MATLAB_MEX_FILE` necessary? I can't find it anywhere in source.

aaronzs replied:
I think it's not necessary, I added it because I'm not sure. 
I found it when I using MATLAB `mex` to compile `.cpp` files. I think the default MATLAB setting is:
`CMDLINE100 : cl /c /Zp8 /GR /W3 /EHs /nologo /MD /O2 /Oy- /DNDEBUG /D_CRT_SECURE_NO_DEPRECATE /D_SCL_SECURE_NO_DEPRECATE /D_SECURE_SCL=0   /DMATLAB_MEX_FILE -Iinclude  -I"C:\Program Files\MATLAB\R2015aSP1\extern\include" ... `

Without these parameters we can also get valid `caffe_.mexw64`.

mnidza replied:
@aaronzs OK, than let's keep it.

lunzueta commented:
I've just tested the last version of my repo with the changes you added and now I see that the MEX file generated when compiled in Debug also works correctly :-)

mnidza commented on windows/scripts/MatlabPostBuild.cmd:
```diff
@@ -0,0 +1,12 @@
+
+echo MatlabPostBuild.cmd : copy matlab generated scripts to output.
+
+copy /y "%SOLUTION_DIR%..\matlab\+caffe\*.m" "%OUTPUT_DIR%matcaffe\+caffe"
```
Is it necessary to copy all of this to build folder? I think that `classification_demo` and `hdf5creation` are not needed, as they are just examples. `+caffe\imagenet` seems to be only used by classification demo, so I would drop that as well.

I'm not sure about copying unit tests. We don't do it in Python. Is there a way to execute unit tests easily if only `+caffe\private` is copied to Release?

Also, I notice that `+caffe\private\caffe_.cpp` is also copied; that should definitely be excluded :)

aaronzs replied:
@mnidza @lunzueta If we need a clean release. We can change the prebuild and postbuild scripts to:
```batch
set SOLUTION_DIR=%~1%
set OUTPUT_DIR=%~2%

echo MatlabPreBuild.cmd : Create output directories for matlab scripts.

if not exist "%OUTPUT_DIR%\matcaffe" mkdir "%OUTPUT_DIR%\matcaffe"
if not exist "%OUTPUT_DIR%\matcaffe\+caffe" mkdir "%OUTPUT_DIR%\matcaffe\+caffe"
if not exist "%OUTPUT_DIR%\matcaffe\+caffe\private" mkdir "%OUTPUT_DIR%\matcaffe\+caffe\private"
```

```batch
set SOLUTION_DIR=%~1%
set OUTPUT_DIR=%~2%

echo MatlabPostBuild.cmd : copy matlab generated scripts to output.

@echo run_tests.m > "%temp%\excludelist.txt"
xcopy /y "%SOLUTION_DIR%..\matlab\+caffe\*.m" "%OUTPUT_DIR%matcaffe\+caffe" /exclude:%temp%\excludelist.txt
copy /y "%SOLUTION_DIR%..\matlab\+caffe\private\*.m" "%OUTPUT_DIR%matcaffe\+caffe\private"
move /y "%OUTPUT_DIR%caffe_.*" "%OUTPUT_DIR%matcaffe\+caffe\private"
```

lunzueta replied:
@mnidza OK, I removed the copy of those files, but I'm not sure either about what to do with the test files, because although it's quite simple not to copy the `+caffe\+test` folder, it's not so simple to exclude the `run_tests.m` file from the rest of `m` files of the same folder. According to this (http://stackoverflow.com/questions/4098444/move-all-files-except-some-file-pattern-from-a-dos-command) `xcopy` could be used instead of `copy`, but then a text file should be added which would include only that file to be excluded, which looks a bit strange to me. Wouldn't be good to copy those test files as well? I think they are quite useful to test in Matlab in an easy way if the compilation was good.

lunzueta replied:
@mnidza I just saw now your last response. OK, I'll update the scripts as you say.

lunzueta commented:
I have a doubt about how to configure the Matlab search path to be able to use this, which probably will happen to more people (as can be seen here: https://github.com/rbgirshick/rcnn/issues/56), if we don't explain it clearly in README.md. When I run `classification_demo.m` in the `demo` folder which was at the same level as the folder `+caffe` containing the MEX file, it runs correctly. But, if that demo file is run from another place, even if I include the generated `matcaffe` path in Matlab's search path, it says `Undefined function or variable 'caffe_'.` Any suggestions @aaronzs @mnidza?

aaronzs commented:
@lunzueta On my computer there is no problem just add `<path_to_matcaffe>` (in a different directory) to Matlab path and run `classification_demo.m`. I think you got the error message is because line 58 in `classification_demo.m`. If there is a `+caffe` in parent directory, the script will add it and the priority is higher than your `<path_to_matcaffe>`.

lunzueta commented:
@aaronzs Yes, you are right! I just commented lines 58-62 and it works now. So, then I'd say that the explanation given in README.md is good enough.

aaronzs commented on matlab/+caffe/private/caffe_.cpp:
```diff
@@ -45,7 +45,11 @@ void mxCHECK_FILE_EXIST(const char* file) {
 static vector<shared_ptr<Solver<float> > > solvers_;
 static vector<shared_ptr<Net<float> > > nets_;
 // init_key is generated at the beginning and everytime you call reset
-static double init_key = static_cast<double>(caffe_rng_rand());
```
@mnidza Sorry about the typo...
It seems that MATLAB halt on line 112 in `common.cpp` when `cublasCreate(&cublas_handle_)` is called. Set `CPU_ONLY` will get rid of this part, so CPU-only build have no issues.

mnidza replied:
Can you use `_MSC_VER` to ensure that the modification is enabled only on Windows?

aaronzs replied:
@mnidza Sure. Shall we use `_MSC_VER` or `_WIN32`?

mnidza commented on windows/matcaffe/matcaffe.vcxproj:
```diff
@@ -0,0 +1,131 @@
+    <TargetName>caffe_</TargetName>
+  </PropertyGroup>
+  <PropertyGroup>
+    <IncludePath>$(MatlabDir)\extern\include;$(IncludePath)</IncludePath>
```
You don't need this since you're importing CommonSettings.props.

lunzueta commented:
@aaronzs with the last update I'm getting the following compilation error in VS2013, which I didn't get in a previous version:
`2>  caffe_.cpp`
`2>..\..\matlab\+caffe\private\caffe_.cpp(16): fatal error C1083: Cannot open include file: 'mex.h': No such file or directory`
What happened?

aaronzs commented:
@lunzueta Do you use the latest `CommonSettings.props`?
because `<IncludePath>$(MatlabDir)\extern\include;$(IncludePath)</IncludePath>` is already in `CommonSettings.props`. I removed it from `matcaffe.matcaffe.vcxproj`.

lunzueta commented:
@aaronzs OK, I didn't update that file with the newer version. It works fine now. Thanks.

pavlejosipovic commented:
Ok seems like all issue have been resolved, let's squash all the commits into a single one.
Write the commit message following guide from here http://chris.beams.io/posts/git-commit/.
